lib/charms/tls_certificates_interface/v1/tls_certificates.py

```diff
@@ -0,0 +1,731 @@
+        )
+        logger.info("Certificate revocation sent to provider")
+
+    @staticmethod
+    def _relation_data_is_valid(certificates_data: dict) -> bool:
+        """Checks whether relation data is valid based on json schema.
+
+        Args:
+            certificates_data: Certificate data in dict format.
+
+        Returns:
+            bool: Whether relation data is valid.
+        """
+        try:
+            validate(instance=certificates_data, schema=PROVIDER_JSON_SCHEMA)
+            return True
+        except exceptions.ValidationError:
+            return False
+
+    def _on_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handler triggerred on relation changed events.
+
+        Args:
+            event: Juju event
+
+        Returns:
+            None
+        """
+        relation_data = _load_relation_data(event.relation.data[event.unit])
+        if not self._relation_data_is_valid(relation_data):
+            logger.warning("Relation data did not pass JSON Schema validation")
+            return
+        for certificate in relation_data["certificates"]:
+            self.on.certificate_available.emit(
+                certificate_signing_request=certificate["certificate_signing_request"],
+                certificate=certificate["certificate"],
+                ca=certificate["ca"],
+                chain=certificate["chain"],
+            )
+
+    def _on_update_status(self, event: UpdateStatusEvent) -> None:
+        """Triggered on update status event.
+
+        Goes through each certificate in the "certificates" relation and checks their expiry date.
+        If they are close to expire (<7 days), emits a CertificateAlmostExpiredEvent event and if
+        they are expired, emits a CertificateExpiredEvent.
+
+        Args:
+            event (UpdateStatusEvent): Juju event
+
+        Returns:
+            None
+        """
+        relation = self.model.get_relation("certificates")
+        if not relation:
+            return
+        for unit in relation.units:
+            relation_data = _load_relation_data(relation.data[unit])
+            if self._relation_data_is_valid(relation_data):
+                certificates = relation_data.get("certificates")
+                if not certificates:
+                    continue
+                for certificate_dict in certificates:
+                    certificate = certificate_dict["certificate"]
+                    certificate_object = x509.load_pem_x509_certificate(data=certificate.encode())
+                    time_difference = certificate_object.not_valid_after - datetime.now()
+                    if time_difference.days < 0:
+                        logger.warning("Certificate is expired")
+                        self.on.certificate_expired.emit(certificate=certificate)
+                        self.revoke_certificate(certificate)
+                        continue
+                    if time_difference.days < 7:
```

Comment:
It depends on the certificate use case.
for certificates issued for mTLS - the good expiration default is only 24 hours.
e.g. https://linkerd.io/2.11/features/automatic-mtls/#how-does-linkerd-s-mtls-implementation-work
or https://support.f5.com/csp/article/K16530333

I think `time_difference` can be controlled via parameter e.g. `TLSCertificatesRequires(self, "certificates", "12h")`

Reply:
Agreed, I added a parameter for this